src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,330 @@
+                    reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.SILENT);
+                } else {
+                    // read from local file (this only makes sense for the direct runner)
+                    reader = SamReaderFactory.make().validationStringency(ValidationStringency.SILENT).open(new File(bam));
+                }
+            }
+
+            @Override
+            public void processElement(ProcessContext c) throws Exception {
+                ContextShard shard = c.element();
+                ArrayList<SimpleInterval> ints =new ArrayList<>();
+                ints.add(shard.interval);
+                List<SimpleInterval> subshards = ShardedReadsDatasource.cutToShards(ints, outputShardSize);
+                int currentSubShardIndex = 0;
+                final int lastValidPos = shard.interval.getEnd() + margin;
+                final int firstValidPos = shard.interval.getStart() - margin;
+                SimpleInterval currentSubShard = subshards.get(currentSubShardIndex);
+
+                try (SAMRecordIterator query = reader.queryOverlapping(shard.interval.getContig(), shard.interval.getStart(), shard.interval.getEnd())) {
+                    ArrayList<GATKRead> readsSoFar = new ArrayList<>();
+
+                    while (query.hasNext()) {
+                        SAMRecord r = query.next();
+                        if (!accept(r, shard.interval)) continue;
+                        if (r.getReadUnmappedFlag()) {
+                            // error out if we accept a read that sticks out too far
+                            // (margin was too tight, the shard may end up missing relevant variants)
+                            if (r.getAlignmentEnd()>lastValidPos) {
+                                throw new GATKException("Margin was too tight, a read sticks out by "+(r.getAlignmentEnd()-shard.interval.getEnd())+", going all the way to "+r.getAlignmentEnd());
+                            }
+                            if (r.getAlignmentStart()<firstValidPos) {
+                                throw new GATKException("Margin was too tight, a read starts early by "+(shard.interval.getStart()-r.getAlignmentStart())+", starting at "+r.getAlignmentStart());
+                            }
+                        }
+
+                        while (currentSubShard.getEnd() < r.getStart()) {
+                            if (!readsSoFar.isEmpty()) {
+                                // ship this one.
+                                ContextShard ret = shard.split(currentSubShard).withReads(readsSoFar);
+                                c.output(ret);
+                                readsSoFar = new ArrayList<>();
+                            }
+                            // move to the next shard
+                            currentSubShard = subshards.get(++currentSubShardIndex);
```

Comment:
that should never happen, since the shard only contains read that start inside it. If this went out of bounds it would indicate a bug in the code, and so a stack trace is the right thing to send.